### DISCUSSION TITLE
Custom Storefront, Replacing Existing Liquid Theme on Custom Domain, with Proper Checkout Links

### DISCUSSION THREAD

**Discussion by 1b4Pi** [Shopify/storefront-api-feedback#114] [Help]
I've made a custom storefront to replace our existing liquid theme as our online shopping experience. This custom storefront has a cart page with a checkout link to complete purchases using the cart object (https://shopify.dev/api/liquid/objects/cart).

Everything works fine during development on localhost and on a staging link with a temporary domain. But once we pointed our primary domain to the custom storefront app, our checkout link returns a 404 error. We can only get the checkout link to work by turning off domain forwarding in our Shopify settings and using our my-shopify.com URL as the primary. Making that change gives us a checkout link with our my-shopify.com URL, which makes the link work and not return a 404. However, now our public-facing domain is disconnected from Shopify, so traffic from search engines are not visiting our new custom storefront on our public-facing URL, they are instead visiting our old liquid theme on our my-shopify.com URL.

I've searched on the Shopify Dev Discord, the Storefront API forum, and this Discussion channel. The potential remedy I can find is the following bullet-point:

> How should I set up a custom domain?
> 
> If you are having issues with your .myshopify.com URL being surfaced at checkout, please check that you are initializing the client and API with the custom domain such as https://backyardbotanicgoods.com/api/2020-04/graphql vs. coolstore.myshopify.com/api/2020-04/graphql

I cannot get that endpoint update to work. Changing the endpoint on the client from "https://my-shopify-url.myshopify.com/api/2022-04/graphql.json" to our custom domain (https://our-public-domain.com/api/2022-04/graphql.json" returns a 404 status:

`ClientError: GraphQL Error (Code: 404): {"response":{"error":"<!DOCTYPE html>\n<html lang=\"en\">\n<head>\n<meta charset=\"utf-8\">\n<title>Error</title>\n</head>\n<body>\n<pre>Cannot POST /api/2022-04/graphql.json</pre>\n</body>\n</html>\n","status":404,...`

We're using prisma-labs/graphql-request. 

What am I missing? How do we point our new custom storefront to our public-facing URL, while also allowing visitors to complete checkout on the Shopify side, just like our old liquid theme did?

Many thanks for the time and help, it is very much appreciated.

**Comment by corygibbons** [comment]
@1b4Pi If I'm understanding you correctly I don't believe they can both share the root domain like that, at least I haven't seen it. I use a subdomain for the Shopify portion and the root for the custom storefront (not hosted by Shopify):

- https://our-public-domain.com → Front end customers visit
- https://checkout.our-public-domain.com → Shopify hosted checkout _(as well as API endpoint: https://checkout.our-public-domain.com/api/2022-04/graphql.json)_

**Reply by slavanossar** [reply]
I'm also dealing with this issue, using a subdomain is nice but still poses problems:
* URLs for CTA links in email notifications link to the subdomain
* the logo in the Checkout links back to the subdomain too

Are there any solutions for this? It makes creating a custom store using Storefront API quite difficult.

**Reply by slavanossar** [reply]
This is my hacky solution for anyone else facing the same issue, included in the `<head>` of `theme.liquid`

```html
<script>
  // REDIRECT HACK
  if (window.location.pathname.indexOf('checkout') === -1) { 
    window.location.href = 'https://customdomain.com' + window.location.pathname
  }
</script>
```

Redirects all pages from `checkout.customdomain.com` to `customdomain.com`, except for any Checkout routes (allowing native Shopify checkout experience to continue to work). Using this method requires matching Shopify page URLs within the custom web app (e.g. `/product/<handle>` and `/collections/<handle>`). Our client is also using Klaviyo for things like cart abandonment notifications, so this allows product links within those emails to still function correctly.

Feels dirty, but does the job.
